1. Посмотреть размер файла, если нужно переименовать столбцы, применить стандартную функцию для просмотра информации о датасете

2. Cоздать столбцы с месяцем и днем (наименование дня недели)

3.  Посчитать количество транзакций по месяцам

4.  Посчитать количество транзакций по дням недели в разрезе product_category 

5. Посчитать количество транзакций в зависимости store_location и product_category (pivot table)

6. Посчитать total_revenue (то есть добавить такой столбец)

7. Посчитать  суммарный total_revenue в разрезе store_location и в разрезе product_category (pivot_table)

8. Вывести самый дешевый и дорогой product_type по unit_price в разрезе product_category

9. Посчитать total revenue по месяцам 

10. Найти второй по популярности продукт по transaction_qty в разрезе product_category (то есть мы ищем второй по популярности product type в разрезе каждого product_category по количеству transaction_qty)

In [2]:
#загрузка библиотеки
import pandas as pd
import datetime
import numpy as np

C:\Users\Main_Omel_PC\AppData\Local\Temp\ipykernel_453496\1457559627.py:2: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


 1.Посмотреть размер файла, если нужно переименовать столбцы, применить стандартную функцию для просмотра информации о датасете

In [3]:
# Смотрим размер файла, применяя стандартную функцию для просмотра информации о датасете

In [4]:
def first_check(dataset):
    df = pd.read_excel(dataset)
 
    print()
    print('Первые 3 строки таблицы')
    display(df.head(3))
 
    print()
    print('Последние 3 строки таблицы')
    display(df.tail(3))
 
    print()
    print('Информация о таблице')   
    print(df.info())
 
    print()
    print('Характеристики чисел в таблице')  
    display(df.describe())
 
    print()
    print('Наименование колонок')  
    print(df.columns)

In [5]:
first_check('Coffee Shop Sales.xlsx')


Первые 3 строки таблицы


transaction_id transaction_date transaction_time  transaction_qty  \
0               1       2023-01-01         07:06:11                2   
1               2       2023-01-01         07:08:56                2   
2               3       2023-01-01         07:14:04                2   

   store_id   store_location  product_id  unit_price    product_category  \
0         5  Lower Manhattan          32         3.0              Coffee   
1         5  Lower Manhattan          57         3.1                 Tea   
2         5  Lower Manhattan          59         4.5  Drinking Chocolate   

            product_type            product_detail  
0  Gourmet brewed coffee               Ethiopia Rg  
1        Brewed Chai tea  Spicy Eye Opener Chai Lg  
2          Hot chocolate         Dark chocolate Lg


Последние 3 строки таблицы


transaction_id transaction_date transaction_time  transaction_qty  \
149113          149454       2023-06-30         20:31:34                1   
149114          149455       2023-06-30         20:57:19                1   
149115          149456       2023-06-30         20:57:19                2   

        store_id  store_location  product_id  unit_price product_category  \
149113         8  Hell's Kitchen          45        3.00              Tea   
149114         8  Hell's Kitchen          40        3.75           Coffee   
149115         8  Hell's Kitchen          64        0.80         Flavours   

             product_type  product_detail  
149113  Brewed herbal tea   Peppermint Lg  
149114   Barista Espresso      Cappuccino  
149115      Regular syrup  Hazelnut syrup


Информация о таблице
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149116 entries, 0 to 149115
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   transaction_id    149116 non-null  int64         
 1   transaction_date  149116 non-null  datetime64[ns]
 2   transaction_time  149116 non-null  object        
 3   transaction_qty   149116 non-null  int64         
 4   store_id          149116 non-null  int64         
 5   store_location    149116 non-null  object        
 6   product_id        149116 non-null  int64         
 7   unit_price        149116 non-null  float64       
 8   product_category  149116 non-null  object        
 9   product_type      149116 non-null  object        
 10  product_detail    149116 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(4), object(5)
memory usage: 12.5+ MB
None

Характеристики чисел в таблице


transaction_id               transaction_date  transaction_qty  \
count   149116.000000                         149116    149116.000000   
mean     74737.371872  2023-04-15 11:50:32.173609984         1.438276   
min          1.000000            2023-01-01 00:00:00         1.000000   
25%      37335.750000            2023-03-06 00:00:00         1.000000   
50%      74727.500000            2023-04-24 00:00:00         1.000000   
75%     112094.250000            2023-05-30 00:00:00         2.000000   
max     149456.000000            2023-06-30 00:00:00         8.000000   
std      43153.600016                            NaN         0.542509   

            store_id     product_id     unit_price  
count  149116.000000  149116.000000  149116.000000  
mean        5.342063      47.918607       3.382219  
min         3.000000       1.000000       0.800000  
25%         3.000000      33.000000       2.500000  
50%         5.000000      47.000000       3.000000  
75%         8.000000      60.000000       3.750000  
max         8.000000      87.000000      45.000000  
std         2.074241      17.930020       2.658723


Наименование колонок
Index(['transaction_id', 'transaction_date', 'transaction_time',
       'transaction_qty', 'store_id', 'store_location', 'product_id',
       'unit_price', 'product_category', 'product_type', 'product_detail'],
      dtype='object')


In [9]:
df = pd.read_excel('Coffee Shop Sales.xlsx')

In [10]:
# переименование столбцов (в нашем датасете в этом нет необходимости)
df = df.rename(columns=lambda c: c.lower().replace('-', '_').replace('/', '_').replace(' ', '_'))

2. Cоздать столбцы с месяцем и днем (наименование дня недели)

In [12]:
# Cоздаeм столбец с месяцем
df['month'] = df.transaction_date.dt.month 


In [13]:
# Cоздаeм столбец с днем (наименованием дня недели)
df['weekday'] = df.transaction_date.dt.day_name()
df.head()

transaction_id transaction_date transaction_time  transaction_qty  \
0               1       2023-01-01         07:06:11                2   
1               2       2023-01-01         07:08:56                2   
2               3       2023-01-01         07:14:04                2   
3               4       2023-01-01         07:20:24                1   
4               5       2023-01-01         07:22:41                2   

   store_id   store_location  product_id  unit_price    product_category  \
0         5  Lower Manhattan          32         3.0              Coffee   
1         5  Lower Manhattan          57         3.1                 Tea   
2         5  Lower Manhattan          59         4.5  Drinking Chocolate   
3         5  Lower Manhattan          22         2.0              Coffee   
4         5  Lower Manhattan          57         3.1                 Tea   

            product_type               product_detail  month weekday  
0  Gourmet brewed coffee                  Ethiopia Rg      1  Sunday  
1        Brewed Chai tea     Spicy Eye Opener Chai Lg      1  Sunday  
2          Hot chocolate            Dark chocolate Lg      1  Sunday  
3            Drip coffee  Our Old Time Diner Blend Sm      1  Sunday  
4        Brewed Chai tea     Spicy Eye Opener Chai Lg      1  Sunday

3.  Посчитать количество транзакций по месяцам

In [14]:
# выводим количество транзакций по месяцам
df.groupby('month')['transaction_qty'].sum()

month
1    24870
2    23550
3    30406
4    36469
5    48233
6    50942
Name: transaction_qty, dtype: int64

 Посчитать количество транзакций по дням недели в разрезе product_category 

In [15]:
# выводим количество транзакций по дням недели в разрезе product_category 
pd.crosstab(df.weekday,
            df.product_category,
            values=df.transaction_qty,
            aggfunc='sum',
            margins=True,
            margins_name="Total")


product_category  Bakery  Branded  Coffee  Coffee beans  Drinking Chocolate  \
weekday                                                                       
Friday              3349      120   13099           266                2443   
Monday              3458       96   12995           259                2604   
Saturday            3199      105   12285           266                2431   
Sunday              3250      121   12597           257                2443   
Thursday            3352       97   12902           231                2631   
Tuesday             3253      115   12699           274                2442   
Wednesday           3353      122   12673           275                2463   
Total              23214      776   89250          1828               17457   

product_category  Flavours  Loose Tea  Packaged Chocolate    Tea   Total  
weekday                                                                   
Friday                1579        182                  66  10103   31207  
Monday                1444        161                  61  10153   31231  
Saturday              1520        160                  62   9586   29614  
Sunday                1483        177                  80   9774   30182  
Thursday              1343        175                  77  10354   31162  
Tuesday               1615        178                  70   9803   30449  
Wednesday             1527        177                  71   9964   30625  
Total                10511       1210                 487  69737  214470

In [16]:
# Cоздаeм столбец и посчитаем total_revenue
df['total_revenue'] = df.transaction_qty*df.unit_price
df.head()

transaction_id transaction_date transaction_time  transaction_qty  \
0               1       2023-01-01         07:06:11                2   
1               2       2023-01-01         07:08:56                2   
2               3       2023-01-01         07:14:04                2   
3               4       2023-01-01         07:20:24                1   
4               5       2023-01-01         07:22:41                2   

   store_id   store_location  product_id  unit_price    product_category  \
0         5  Lower Manhattan          32         3.0              Coffee   
1         5  Lower Manhattan          57         3.1                 Tea   
2         5  Lower Manhattan          59         4.5  Drinking Chocolate   
3         5  Lower Manhattan          22         2.0              Coffee   
4         5  Lower Manhattan          57         3.1                 Tea   

            product_type               product_detail  month weekday  \
0  Gourmet brewed coffee                  Ethiopia Rg      1  Sunday   
1        Brewed Chai tea     Spicy Eye Opener Chai Lg      1  Sunday   
2          Hot chocolate            Dark chocolate Lg      1  Sunday   
3            Drip coffee  Our Old Time Diner Blend Sm      1  Sunday   
4        Brewed Chai tea     Spicy Eye Opener Chai Lg      1  Sunday   

   total_revenue  
0            6.0  
1            6.2  
2            9.0  
3            2.0  
4            6.2

<span style="font-weight: bold;"><span style="font-size: 30px;"><span style="background-color: lightpink;">Строим корреляционную матрицу</span></span></span>

In [17]:
corr_df = df[['transaction_id', 'transaction_qty', 'store_id', 'product_id', 'unit_price', 'month', 'total_revenue']]
corr_df

transaction_id  transaction_qty  store_id  product_id  unit_price  \
0                    1                2         5          32        3.00   
1                    2                2         5          57        3.10   
2                    3                2         5          59        4.50   
3                    4                1         5          22        2.00   
4                    5                2         5          57        3.10   
...                ...              ...       ...         ...         ...   
149111          149452                2         8          44        2.50   
149112          149453                2         8          49        3.00   
149113          149454                1         8          45        3.00   
149114          149455                1         8          40        3.75   
149115          149456                2         8          64        0.80   

        month  total_revenue  
0           1           6.00  
1           1           6.20  
2           1           9.00  
3           1           2.00  
4           1           6.20  
...       ...            ...  
149111      6           5.00  
149112      6           6.00  
149113      6           3.00  
149114      6           3.75  
149115      6           1.60  

[149116 rows x 7 columns]

In [18]:
corr_df.corr()

transaction_id  transaction_qty  store_id  product_id  \
transaction_id         1.000000         0.010250 -0.006054    0.001076   
transaction_qty        0.010250         1.000000 -0.000794   -0.175255   
store_id              -0.006054        -0.000794  1.000000    0.025727   
product_id             0.001076        -0.175255  0.025727    1.000000   
unit_price             0.000836        -0.123546  0.001891   -0.140423   
month                  0.972616         0.002849 -0.000900   -0.002370   
total_revenue          0.005949         0.356231  0.004595   -0.167461   

                 unit_price     month  total_revenue  
transaction_id     0.000836  0.972616       0.005949  
transaction_qty   -0.123546  0.002849       0.356231  
store_id           0.001891 -0.000900       0.004595  
product_id        -0.140423 -0.002370      -0.167461  
unit_price         1.000000 -0.002004       0.685550  
month             -0.002004  1.000000       0.001243  
total_revenue      0.685550  0.001243       1.000000

Вывод

Фактор, который в большей степени связан с total_revenue: unit_price(КК = 0.685550) 

In [19]:
import statsmodels.formula.api as smf 
import statsmodels.api as sm

In [20]:
X = corr_df.unit_price
Y = corr_df.total_revenue

In [21]:
X = sm.add_constant(X)  # добавить константу, чтобы был свободный член
model = sm.OLS(Y, X)  # говорим модели, что у нас ЗП, а что НП зависимая переменная и независимая переменная
results = model.fit()  # строим регрессионную прямую
print(results.summary())  # смотрим результат
#построение регрессионной модели

                            OLS Regression Results                            
Dep. Variable:          total_revenue   R-squared:                       0.470
Model:                            OLS   Adj. R-squared:                  0.470
Method:                 Least Squares   F-statistic:                 1.322e+05
Date:                Sun, 07 Apr 2024   Prob (F-statistic):               0.00
Time:                        16:48:09   Log-Likelihood:            -3.7921e+05
No. Observations:              149116   AIC:                         7.584e+05
Df Residuals:                  149114   BIC:                         7.584e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.9999      0.013     77.539      0.0

<span style="background-color: lightblue;">Коэффициент детерминации =  0.470</span>

<span style="background-color: lightblue;">Коэффициент корреляции = 0.69</span>

<span style="background-color: lightblue;">у = 0.99 + 1.09х</span>

<span style="background-color: lightpink;">При увеличении unit_price на 1 total_revenue увеличивается на 1.09</span>

In [ ]:
sns.regplot(x = dfunit_price, y = df.price, data = df)